expression/builtin_like_vec.go

```diff
@@ -27,17 +30,80 @@ func (b *builtinLikeSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) er
 }
 
 func (b *builtinRegexpBinarySig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinRegexpBinarySig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	return vecEvalRegexp(&b.baseBuiltinFunc, b, input, result)
 }
 
 func (b *builtinRegexpSig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinRegexpSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	return vecEvalRegexp(&b.baseBuiltinFunc, b, input, result)
+}
+
+func vecEvalRegexp(b *baseBuiltinFunc, rc regexpCompiler, input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+	bufExpr, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(bufExpr)
+	if err := b.args[0].VecEvalString(b.ctx, input, bufExpr); err != nil {
+		return err
+	}
+
+	bufPat, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(bufPat)
+	if err := b.args[1].VecEvalString(b.ctx, input, bufPat); err != nil {
+		return err
+	}
+
+	pat2re := make(map[string]*regexp.Regexp)
+	isMemoizable := b.args[1].ConstItem()
+	if isMemoizable {
+		// If args[1] is a constant item, then all the items in bufPat are
+		// the same. So the pattern can be compiled once and memoized.
+		for i := 0; i < n; i++ {
+			if bufPat.IsNull(i) {
+				continue
+			}
+			pat := bufPat.GetString(i)
+			if _, ok := pat2re[pat]; ok {
+				continue
+			}
+			re, err := rc.compile(pat)
+			if err != nil {
+				return err
+			}
+			pat2re[pat] = re
+		}
+	}
```

Comment:
Does this memorization only work during this invoke? In fact, this function is called for each block of tuples, so it is better to memorize the `compile` results for this function.

Reply:
Yes, it currently only memoizes per invocation. To achieve what you have suggested, I need some clarification:

* Is the same `builtinRegexpSig` instance used for all the chunks from the same query? 
* Does TiDB instantiate a new `builtinRegexpSig instance for every new query?

If so, maybe we can just add a memoized field to the `builtinRegexpSig` struct itself?

Also, are these chunks invoked sequentially or in parallel?

Reply:
Yes for the two querstions. We build `XXXSig` for each query.

It is an option to add a memorized field, or may need a new `XXXSig` aiming at constant cases.

Each thread sequentially invokes the functions, but multiple threads run in parallel. But each thread clone a local instance of the expression.

Reply:
> Each thread sequentially invokes the functions, but multiple threads run in parallel. But each thread clone a local instance of the expression.

Sorry, I guess my real question is, will the same `builtinRegexpSig` be used in multiple goroutines simultaneously?

Reply:
I've updated the code, assuming that there is no concurrency issue. 

I'm not sure if the updated code is a better solution. Memoizing the regex across different function calls didn't improve the performance a lot, but did make the code more complicated :-/

Reply:
>Sorry, I guess my real question is, will the same builtinRegexpSig be used in multiple goroutines simultaneously?

No. Each goroutine owns a copy of the instance of `builtinRegexpSig `


> I've updated the code, assuming that there is no concurrency issue.

> I'm not sure if the updated code is a better solution. Memoizing the regex across different function calls didn't improve the performance a lot, but did make the code more complicated :-/

The solutions is nice. 
The improvement may be obvious if the function is called millions of times in real query execution.

Reply:
Thanks, ptal!